### PR TITLE
Replacement card for vertical_button + vertical_button_custom_state

### PR DESCRIPTION
**Pull request by mp-se**
Adress feature request #191 , enable vertical button to be used as switch with input_booelan. Everything is now in one card which can be extended to handle other entites if needed. 

Tested the card with both input_boolean and input_select entities. 

I still have an issue with backwards compatibility parts for vertical_button_custom_state which i cant find the solution to. Need help on that one. More info in the feature request.

### CODE REVIEW

**Comment by mp-se** [comment]
It looks like my problem is connected to this, https://github.com/custom-cards/button-card/issues/506

**Comment by mp-se** [comment]
There are apparently limitation what you can nest in templates. I had to put the code that handles the old variable names in the base template. not the way i would like to have it. but it does not work otherwise. i put a comment in there so when the old templates are removed we can remove that code as well.

File: config/minimalist-templates/button_card_templates.yaml [hunk 1]
```diff
@@ -275,10 +275,34 @@ card_title:
...
+          if(typeof variables.color !== 'undefined')
+            return variables.color;
+         return "blue";
+      ]]]
+    ulm_card_vertical_button_state: |
+      [[[
+        if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
+          if(typeof variables.color !== 'undefined')
+            return variables.state;
+         return "on";
```
Comment (saxel) [inline config/minimalist-templates/button_card_templates.yaml:294]: Indentation error
```suggestion
        return "on";
```

File: config/minimalist-templates/button_card_templates.yaml [hunk 2]
```diff
@@ -275,10 +275,34 @@ card_title:
...
+      [[[
+        if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
+          if(typeof variables.color !== 'undefined')
+            return variables.color;
+         return "blue";
+      ]]]
+    ulm_card_vertical_button_state: |
+      [[[
+        if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
+          if(typeof variables.color !== 'undefined')
```
Comment (saxel) [inline config/minimalist-templates/button_card_templates.yaml:292]:
```suggestion
          if(typeof variables.state !== 'undefined')
```

File: config/minimalist-templates/button_card_templates.yaml [hunk 3]
```diff
@@ -275,10 +275,34 @@ card_title:
...
 ### VERTICAL BUTTONS (fka SCENES) ###
-vertical_buttons:
+card_vertical_button:
+  variables:
+    ulm_card_vertical_button_color: |
+      [[[
+        if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
+          if(typeof variables.color !== 'undefined')
+            return variables.color;
+         return "blue";
```
Comment (saxel) [inline config/minimalist-templates/button_card_templates.yaml:287]:
```suggestion
        return "blue";
```

**Review by saxel — COMMENTED** [review]
I tried your changes and it doesn't seem like the compatability mode works still?

image

If it is not possible to solve then I think there is no reason to keep the custom_state template.

**Comment by mp-se** [comment]
Let me check, if I missed something copying the changes from my HA directory.

**Comment by mp-se** [comment]
i've tested the code on my HA now and for me it works. Perhaps I have used the orginal cards diffrently, this is my test configs:
```
- type: 'custom:button-card'
  template: card_vertical_button
  name: Kitchen
  entity: input_boolean.scene_kitchen_work

- type: 'custom:button-card'
  template: card_vertical_button
  entity: input_select.test_vertical_buttons
  name: Away
  icon: mdi:television-classic
  variables:
    ulm_card_vertical_button_state: Away

- type: 'custom:button-card'
  template: card_vertical_button
  entity: input_select.test_vertical_buttons
  name: Night
  icon: mdi:television-classic
  variables:
    ulm_card_vertical_button_state: Night
    ulm_card_vertical_button_color: green

- type: 'custom:button-card'
  template: vertical_buttons
  name: Kitchen (legacy)
  entity:  input_boolean.scene_kitchen_work

- type: 'custom:button-card'
  template: vertical_buttons_custom_state
  entity: input_select.test_vertical_buttons
  name: Night2 (legacy)
  variables:
    state: "Night"
    color: "red"

- type: 'custom:button-card'
  template: _vertical_buttons_custom_state
  entity: input_select.test_vertical_buttons
  name: Night2 (legacy olg)
  variables:
    state: "Night"
    color: "red"

- type: 'custom:button-card'
  template: vertical_buttons_custom_state
  entity: input_select.test_vertical_buttons
  name: Night2 (legacy)
  variables:
    state: "Night"

- type: 'custom:button-card'
  template: _vertical_buttons_custom_state
  entity: input_select.test_vertical_buttons
  name: Night2 (legacy olg)
  variables:
    state: "Night"
```

**Comment by saxel** [comment]
Ah, yeah I used the variable.state to display as the name as well. But that's fine. Does activating the legacy button work for you?

**Comment by mp-se** [comment]
> Ah, yeah I used the variable.state to display as the name as well. But that's fine. Does activating the legacy button work for you?

No, it doesent do anything, i guess that was a thing paddy was going to adress.

**Comment by CM000n** [comment]
what do you guys think, is this ready to merge, or rather in a WIP status?

**Comment by mp-se** [comment]
> what do you guys think, is this ready to merge, or rather in a WIP status?

According to how i interpret the design, it works as before so it can be merged.  I have not found any docs but if anyone has other views i can test/update according to that input.

**Comment by saxel** [comment]
Yeah the updated card works, but not the compatibility ones. All it results in is the "dummy" buttons that doesn't work, no color on active state, and no actions happen when pressed.

You have code for compatibility mode in the updated card, but it doesn't work. the variables are always undefined. Is it supposed to work?
```yaml
    ulm_card_vertical_button_color: |
      [[[
        if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
          if(typeof variables.color !== 'undefined')
            return variables.color;
         return "blue";
      ]]]
    ulm_card_vertical_button_state: |
      [[[
        if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
          if(typeof variables.color !== 'undefined')
            return variables.state;
         return "on";
      ]]]
```

**Comment by mp-se** [comment]
> Yeah the updated card works, but not the compatibility ones. All it results in is the "dummy" buttons that doesn't work, no color on active state, and no actions happen when pressed.
> 
> You have code for compatibility mode in the updated card, but it doesn't work. the variables are always undefined. Is it supposed to work?
> 
> ```yaml
>     ulm_card_vertical_button_color: |
>       [[[
>         if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
>           if(typeof variables.color !== 'undefined')
>             return variables.color;
>          return "blue";
>       ]]]
>     ulm_card_vertical_button_state: |
>       [[[
>         if(typeof variables !== 'undefined' ) // This code is for backwards compatibility
>           if(typeof variables.color !== 'undefined')
>             return variables.state;
>          return "on";
>       ]]]
> ```

It works on my installation, so we might be using them diffrently. Can you share how you use the cards so I can test those scenarios ?

**Comment by saxel** [comment]
I tested with your examples that were using input_select, but switched out the entity and the state value to my own, also tested with name attribute removed. Something weird happens when the page is first loaded. First nothing happens when I click the buttons. Then I click a button which has been has been migrated to card_vertical_button. Suddenly the label of the legacy button changes from "on" to the state value. And now the first two legacy buttons work, but not the last two without color variable. I'm not sure what is going on here.

vertical_buttons_custom_state3

```yaml
              - type: 'custom:button-card'
                template: vertical_buttons_custom_state
                entity: input_select.status_saxel
                name: Night2 (legacy)
                variables:
                  state: "home"
                  color: "red"

              - type: 'custom:button-card'
                template: vertical_buttons_custom_state
                entity: input_select.status_saxel
                variables:
                  state: "away"
                  color: "red"

              - type: 'custom:button-card'
                template: vertical_buttons_custom_state
                entity: input_select.status_saxel
                #name: Night2 (legacy)
                variables:
                  state: "exercise"

              - type: 'custom:button-card'
                template: vertical_buttons_custom_state
                entity: input_select.status_saxel
                name: Night2 (legacy olg)
                variables:
                  state: "work"
```

**Comment by mp-se** [comment]
Wonder if this is connected to that I use service calls to switch states , since it needs to be processed via HA event handling this could cause a delay before the state change is detected. On my system the delay is not visible. What are you running you HA instance on? But if there are issues it might be better to just add the new card and deprecate the other vertical buttons?

**Comment by saxel** [comment]
I don't think there's an issue with delays since the legacy buttons starts working after I press any migrated button. My HA is running in a VM with plenty of resources available so I don't think that is an issue either.

Every time I reload/refresh the page I can reproduce it like this

**Comment by mp-se** [comment]
This is how it looks on my system

vb

**Comment by saxel** [comment]
Super weird, I don't know why it would work for you but not for me 🤷‍♂️I would anyway just migrate straight away, but I was curious about the legacy/compatibility card.

**Comment by mp-se** [comment]
> Super weird, I don't know why it would work for you but not for me 🤷‍♂️I would anyway just migrate straight away, but I was curious about the legacy/compatibility card.

I agree, its not a good sign if it behaves differently that always makes me uncomfortable.  Perhaps I should keep the legacy templates for now and just introduce the new card ?

**Comment by mp-se** [comment]
To support issue #198 there are code changes needed. It currently only supports input_boolean and input_select as data types. Its not difficult to add those but I would like to have a decision on what is the next step regarding backwards compatibility

**Comment by saxel** [comment]
If we absolutely don't want to break compatibility then we need to keep the legacy template for at least 1 release.

**Comment by CM000n** [comment]
Since the next release will probably contain a lot of breaking changes anyway, I think we can make a cut here and go the new way and deprecate the old templates.
If there are issues popping up regarding incompatibility and no longer working vertical buttons, we can then refer them to the new templates.

**Comment by mp-se** [comment]
I will fix this later today then, i will probably align this with the new structure as well. so i might skip this PR and open a new one.

**Comment by mp-se** [comment]
I've done the updates and also added so entities of type light & switch can be handled. That should support #198 if the user adds show_name: false to their defintion

**Pull request merged**